crates/biome_js_analyze/src/utils/restricted_glob.rs

```diff
@@ -0,0 +1,154 @@
+/// - globstar `**` that matches zero or more path segments
+/// - Use `\*` to escape `*`
+/// - `?`, `[`, `]`, `{`, and `}` must be escaped using `\`.
+///   These characters are reserved for future use.
+/// - `!` must be escaped if it is the first characrter of the pattern
+///
+/// A path segment is delimited by path separator `/` or the start/end of the path.
+#[derive(Clone, Debug, Deserializable, serde::Deserialize, serde::Serialize)]
+#[serde(try_from = "String", into = "String")]
+pub struct RestrictedGlob(globset::GlobMatcher);
+
+impl std::ops::Deref for RestrictedGlob {
+    type Target = globset::GlobMatcher;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl std::fmt::Display for RestrictedGlob {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let repr = self.0.glob().to_string();
+        f.write_str(&repr)
+    }
+}
+
+impl From<RestrictedGlob> for String {
+    fn from(value: RestrictedGlob) -> Self {
+        value.to_string()
+    }
+}
+
+impl std::str::FromStr for RestrictedGlob {
+    type Err = globset::ErrorKind;
+
+    fn from_str(value: &str) -> Result<Self, Self::Err> {
+        is_restricted_glob(value)?;
+        let mut glob_builder = globset::GlobBuilder::new(value);
+        // Allow escaping with `\` on all platforms.
+        glob_builder.backslash_escape(true);
+        // Only `**` can match `/`
+        glob_builder.literal_separator(true);
+        match glob_builder.build() {
+            Ok(glob) => Ok(RestrictedGlob(glob.compile_matcher())),
+            Err(error) => Err(error.kind().clone()),
+        }
+    }
+}
+
+impl TryFrom<String> for RestrictedGlob {
+    type Error = globset::ErrorKind;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        value.parse()
+    }
+}
+
+#[cfg(feature = "schemars")]
+impl schemars::JsonSchema for RestrictedGlob {
+    fn schema_name() -> String {
+        "Regex".to_string()
+    }
+
+    fn json_schema(gen: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
+        String::json_schema(gen)
+    }
+}
+
+/// Returns an error if `pattern` doesn't follow the restricted glob syntax.
+fn is_restricted_glob(pattern: &str) -> Result<(), globset::ErrorKind> {
+    let mut it = pattern.bytes().enumerate();
+    while let Some((i, c)) = it.next() {
+        match c {
+            b'!' if i == 0 => {
+                return Err(globset::ErrorKind::Regex(
+                    r"Negated globs `!` are not supported. Use `\!` to escape the character."
+                        .to_string(),
+                ));
+            }
+            b'\\' => {
+                // Accept a restrictive set of escape sequence
+                if let Some((_, c)) = it.next() {
+                    if !matches!(c, b'!' | b'*' | b'?' | b'{' | b'}' | b'[' | b']' | b'\\') {
+                        // SAFETY: safe because of the match
```

Comment:
nit: the phrase "because of the match" is only understandable by someone who knows what's happening inside the match. For example, I genuinely don't know, and I have to venture into a guess, is it because the characters checked can be represented in `u32`?